service/src/main/java/com/epam/esm/service/GiftCertificateService.java

```diff
@@ -0,0 +1,227 @@
+package com.epam.esm.service;
+
+import com.epam.esm.dto.GiftCertificateDto;
+import com.epam.esm.dto.TagDto;
+import com.epam.esm.entity.GiftCertificate;
+import com.epam.esm.entity.Tag;
+import com.epam.esm.exception.EntityNotFoundException;
+import com.epam.esm.exception.InvalidEntityException;
+import com.epam.esm.exception.InvalidSortStringException;
+import com.epam.esm.repository.GiftCertificateRepository;
+import com.epam.esm.repository.TagRepository;
+import com.epam.esm.validator.GiftCertificateValidator;
+import com.epam.esm.validator.SortStringValidator;
+import com.epam.esm.validator.TagValidator;
+import com.epam.esm.validator.ValidationError;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Propagation;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.time.Instant;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * This service class encapsulated business logic related to {@link GiftCertificate} entity.
+ *
+ * @author Evgeniy Sokolchik
+ */
+@Service
+public class GiftCertificateService {
+
+  private final GiftCertificateRepository giftCertificateRepository;
+  private final TagRepository tagRepository;
+  private final GiftCertificateValidator giftCertificateValidator;
+  private final TagValidator tagValidator;
+  private final SortStringValidator sortStringValidator;
+
+  public GiftCertificateService(
+      GiftCertificateRepository giftCertificateRepository,
+      TagRepository tagRepository,
+      GiftCertificateValidator giftCertificateValidator,
+      TagValidator tagValidator,
+      SortStringValidator sortStringValidator) {
```

Comment:
consider using @RequiredArgsConstructor as long as you use lombok library

Reply:
done